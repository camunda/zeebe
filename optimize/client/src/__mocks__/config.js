/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export const isEmailEnabled = () => true;
export const isSharingEnabled = () => true;
export const getOptimizeVersion = () => '2.7.0';
export const getWebappEndpoints = () => ({
  default: {endpoint: 'http://localhost:8080/camunda', engineName: 'default'},
});
export const getHeader = () => ({
  textColor: 'dark', // or "light"
  backgroundColor: '#FFFFFF',
  logo: '',
});
export const ConfigProvider = () => <div />;
export {default as newReport} from '../modules/newReport.json';
