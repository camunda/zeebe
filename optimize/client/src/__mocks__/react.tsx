/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

// @ts-ignore
export * from 'react';
export {default} from 'react';

const outstandingEffects: (() => void)[] = [];
const outstandingCleanups: (() => void)[] = [];

export const useEffect = (fn: () => (() => void) | void) => outstandingEffects.push(fn);
export const runLastEffect = () => {
  if (outstandingEffects.length) {
    const cleanup = outstandingEffects.pop()?.();

    if (cleanup) {
      outstandingCleanups.push(cleanup);
    }
  }
};

export const runLastCleanup = () => {
  if (outstandingCleanups.length) {
    outstandingCleanups.pop()?.();
  }
};

export const runAllEffects = () => {
  const numberOfEffects = outstandingEffects.length;

  for (let i = 0; i < numberOfEffects; i++) {
    const cleanup = outstandingEffects.shift()?.();

    if (cleanup) {
      outstandingCleanups.push(cleanup);
    }
  }
};

beforeEach(() => {
  outstandingEffects.length = 0;
});

export const runAllCleanups = () => {
  const numberOfEffects = outstandingCleanups.length;

  for (let i = 0; i < numberOfEffects; i++) {
    outstandingCleanups.shift()?.();
  }
};
