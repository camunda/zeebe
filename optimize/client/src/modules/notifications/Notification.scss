@keyframes appear {
  from {
    opacity: 0;
    margin-top: -60px;
  }
}

@keyframes disappear {
  to {
    opacity: 0;
    margin-top: -60px;
  }
}

.Notification {
  position: relative;
  pointer-events: auto;
  margin: auto;
  width: 731px;
  border-radius: 3px;
  background-color: var(--blue-base-65);
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.5);
  padding: 20px 50px;
  margin-top: 34px;
  color: var(--white);
  animation: appear 0.35s ease-in-out;

  > .Icon {
    position: absolute;
    left: 20px;
    top: 21px;
  }

  .Button.close {
    position: absolute;
    right: 20px;
    top: 20px;
    background: transparent;
    border: none;
    color: var(--white);
    padding: 1px;

    .Icon {
      margin-right: 0;
    }
  }

  &.success {
    background-color: var(--green-base-44);
  }

  &.warning {
    background-color: var(--orange-base-60);
  }

  &.error {
    background-color: var(--red-darken-38);
  }

  &.closing {
    animation: disappear 0.35s ease-in-out forwards;
  }
}
