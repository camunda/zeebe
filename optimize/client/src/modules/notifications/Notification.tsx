/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {useState, useEffect, useRef, useCallback, ReactNode} from 'react';
import classnames from 'classnames';
import {Button, Icon} from 'components';

import './Notification.scss';

export interface Config {
  id?: string;
  type?: string;
  text: ReactNode;
  stayOpen?: boolean;
  duration?: number;
}

interface NotificationProps {
  config: Config;
  remove: () => void;
}

export default function Notification({config, remove}: NotificationProps): JSX.Element {
  const [closing, setClosing] = useState(false);
  const closeTrigger = useRef<ReturnType<typeof setTimeout>>();

  const getIconType = (): string | null => {
    switch (config.type) {
      case 'success':
        return 'check-large';
      case 'warning':
        return 'warning';
      case 'error':
        return 'error';
      case 'hint':
        return 'hint';
      default:
        return null;
    }
  };

  const close = useCallback(() => {
    setClosing(true);
    setTimeout(remove, 350);
  }, [remove]);

  useEffect(() => {
    if (!config.stayOpen) {
      closeTrigger.current = setTimeout(close, config.duration || 4350);
    }
    return () => {
      clearTimeout(closeTrigger.current);
    };
  }, [config, closeTrigger, close]);

  const keepOpen = () => {
    clearTimeout(closeTrigger.current);
  };

  const iconType = getIconType();

  return (
    <div className={classnames('Notification', config.type, {closing})} onClick={keepOpen}>
      {iconType && <Icon type={iconType} />}
      {config.text}
      <Button className="close" onClick={close}>
        <Icon type="close-large" />
      </Button>
    </div>
  );
}
