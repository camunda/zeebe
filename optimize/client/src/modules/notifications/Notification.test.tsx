/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {runLastEffect} from '__mocks__/react';
import {shallow} from 'enzyme';

import Notification from './Notification';

jest.useFakeTimers();

it('should render text provided in config', () => {
  const node = shallow(
    <Notification config={{text: 'Sample', stayOpen: true}} remove={jest.fn()} />
  );

  expect(node).toMatchSnapshot();
});

it('should call the provided remove function after a while', () => {
  const spy = jest.fn();
  shallow(<Notification config={{text: 'Sample'}} remove={spy} />);

  runLastEffect();
  jest.runAllTimers();

  expect(spy).toHaveBeenCalled();
});

it('should not remove the Notification if it should stay open', () => {
  const spy = jest.fn();
  shallow(<Notification config={{text: 'Sample', stayOpen: true}} remove={spy} />);

  runLastEffect();
  jest.runAllTimers();

  expect(spy).not.toHaveBeenCalled();
});

it('should not remove the Notification if the user interacts with it', () => {
  const spy = jest.fn();
  const node = shallow(<Notification config={{text: 'Sample', stayOpen: true}} remove={spy} />);

  node.simulate('click');

  jest.runAllTimers();

  expect(spy).not.toHaveBeenCalled();
});

it('should remove the Notification after the specified delay', () => {
  const spy = jest.fn();
  shallow(<Notification config={{text: 'Sample', duration: 1234}} remove={spy} />);

  runLastEffect();
  jest.advanceTimersByTime(1230);

  expect(spy).not.toHaveBeenCalled();

  jest.advanceTimersByTime(1235);

  expect(spy).toHaveBeenCalled();
});
