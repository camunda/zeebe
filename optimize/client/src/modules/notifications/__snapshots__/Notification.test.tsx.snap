// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render text provided in config 1`] = `
<div
  className="Notification"
  onClick={[Function]}
>
  Sample
  <ForwardRef(Button)
    className="close"
    onClick={[Function]}
  >
    <Icon
      type="close-large"
    />
  </ForwardRef(Button)>
</div>
`;
